File: bci/api/typescript/src/core/SignalProcessor.ts
/**
 * WIA BCI Signal Processor
 * @module wia-bci/core/SignalProcessor
 */

import type { BandPowers, FrequencyBand, PowerSpectrum, FREQUENCY_BANDS } from '../types';

/**
 * Frequency band ranges in Hz
 */
const BANDS: Record<string, [number, number]> = {
  delta: [0.5, 4],
  theta: [4, 8],
  alpha: [8, 13],
  beta: [13, 30],
  gamma: [30, 100],
};

/**
 * Signal Processing Utilities
 */
export class SignalProcessor {
  /**
   * Apply high-pass filter
   * Uses simple first-order IIR filter
   */
  static highpass(data: Float32Array, cutoff: number, fs: number): Float32Array {
    const result = new Float32Array(data.length);
    const rc = 1 / (2 * Math.PI * cutoff);
    const dt = 1 / fs;
    const alpha = rc / (rc + dt);

    result[0] = data[0];
    for (let i = 1; i < data.length; i++) {
      result[i] = alpha * (result[i - 1] + data[i] - data[i - 1]);
    }

    return result;
  }

  /**
   * Apply low-pass filter
   * Uses simple first-order IIR filter
   */
  static lowpass(data: Float32Array, cutoff: number, fs: number): Float32Array {
    const result = new Float32Array(data.length);
    const rc = 1 / (2 * Math.PI * cutoff);
    const dt = 1 / fs;
    const alpha = dt / (rc + dt);

    result[0] = data[0];
    for (let i = 1; i < data.length; i++) {
      result[i] = result[i - 1] + alpha * (data[i] - result[i - 1]);
    }

    return result;
  }

  /**
   * Apply band-pass filter
   */
  static bandpass(data: Float32Array, low: number, high: number, fs: number): Float32Array {
    const hp = this.highpass(data, low, fs);
    return this.lowpass(hp, high, fs);
  }

  /**
   * Apply notch filter (e.g., for 50/60 Hz line noise)
   */
  static notch(data: Float32Array, freq: number, fs: number, Q = 30): Float32Array {
    const result = new Float32Array(data.length);
    const w0 = (2 * Math.PI * freq) / fs;
    const alpha = Math.sin(w0) / (2 * Q);

    const b0 = 1;
    const b1 = -2 * Math.cos(w0);
    const b2 = 1;
    const a0 = 1 + alpha;
    const a1 = -2 * Math.cos(w0);
    const a2 = 1 - alpha;

    // Normalize
    const nb0 = b0 / a0;
    const nb1 = b1 / a0;
    const nb2 = b2 / a0;
    const na1 = a1 / a0;
    const na2 = a2 / a0;

    // Apply biquad filter
    let x1 = 0, x2 = 0, y1 = 0, y2 = 0;

    for (let i = 0; i < data.length; i++) {
      const x = data[i];
      const y = nb0 * x + nb1 * x1 + nb2 * x2 - na1 * y1 - na2 * y2;

      x2 = x1;
      x1 = x;
      y2 = y1;
      y1 = y;

      result[i] = y;
    }

    return result;
  }

  /**
   * Compute Power Spectral Density using FFT
   */
  static psd(data: Float32Array, fs: number): PowerSpectrum {
    const n = data.length;
    const nfft = Math.pow(2, Math.ceil(Math.log2(n)));

    // Zero-pad and apply Hanning window
    const windowed = new Float32Array(nfft);
    for (let i = 0; i < n; i++) {
      const window = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (n - 1)));
      windowed[i] = data[i] * window;
    }

    // Simple DFT (for production, use proper FFT library)
    const powers = new Float32Array(nfft / 2);
    const frequencies = new Float32Array(nfft / 2);

    for (let k = 0; k < nfft / 2; k++) {
      let real = 0;
      let imag = 0;

      for (let n2 = 0; n2 < nfft; n2++) {
        const angle = (2 * Math.PI * k * n2) / nfft;
        real += windowed[n2] * Math.cos(angle);
        imag -= windowed[n2] * Math.sin(angle);
      }

      powers[k] = (real * real + imag * imag) / nfft;
      frequencies[k] = (k * fs) / nfft;
    }

    return {
      frequencies,
      powers,
      resolution: fs / nfft,
    };
  }

  /**
   * Calculate power in a frequency band
   */
  static bandpower(data: Float32Array, fs: number, band: FrequencyBand): number {
    const [low, high] = typeof band === 'string' ? BANDS[band] : band;

    const spectrum = this.psd(data, fs);
    let power = 0;
    let count = 0;

    for (let i = 0; i < spectrum.frequencies.length; i++) {
      const freq = spectrum.frequencies[i];
      if (freq >= low && freq <= high) {
        power += spectrum.powers[i];
        count++;
      }
    }

    return count > 0 ? power / count : 0;
  }

  /**
   * Calculate all standard frequency band powers
   */
  static allBandPowers(data: Float32Array, fs: number): BandPowers {
    return {
      delta: this.bandpower(data, fs, 'delta'),
      theta: this.bandpower(data, fs, 'theta'),
      alpha: this.bandpower(data, fs, 'alpha'),
      beta: this.bandpower(data, fs, 'beta'),
      gamma: this.bandpower(data, fs, 'gamma'),
    };
  }

  /**
   * Extract epoch from continuous data
   */
  static epoch(data: Float32Array, start: number, end: number): Float32Array {
    const length = end - start;
    const result = new Float32Array(length);

    for (let i = 0; i < length; i++) {
      result[i] = data[start + i] ?? 0;
    }

    return result;
  }

  /**
   * Generate sliding windows
   */
  static sliding(data: Float32Array, windowSize: number, step: number): Float32Array[] {
    const windows: Float32Array[] = [];

    for (let start = 0; start + windowSize <= data.length; start += step) {
      windows.push(this.epoch(data, start, start + windowSize));
    }

    return windows;
  }

  /**
   * Detect eye blinks in EOG channel
   */
  static detectBlinks(eog: Float32Array, threshold = 100): number[] {
    const blinks: number[] = [];
    let inBlink = false;

    for (let i = 0; i < eog.length; i++) {
      if (Math.abs(eog[i]) > threshold && !inBlink) {
        blinks.push(i);
        inBlink = true;
      } else if (Math.abs(eog[i]) < threshold / 2) {
        inBlink = false;
      }
    }

    return blinks;
  }

  /**
   * Calculate RMS (Root Mean Square)
   */
  static rms(data: Float32Array): number {
    let sum = 0;
    for (let i = 0; i < data.length; i++) {
      sum += data[i] * data[i];
    }
    return Math.sqrt(sum / data.length);
  }

  /**
   * Calculate mean
   */
  static mean(data: Float32Array): number {
    let sum = 0;
    for (let i = 0; i < data.length; i++) {
      sum += data[i];
    }
    return sum / data.length;
  }

  /**
   * Calculate variance
   */
  static variance(data: Float32Array): number {
    const m = this.mean(data);
    let sum = 0;
    for (let i = 0; i < data.length; i++) {
      const diff = data[i] - m;
      sum += diff * diff;
    }
    return sum / data.length;
  }

  /**
   * Count zero crossings
   */
  static zeroCrossings(data: Float32Array): number {
    let count = 0;
    for (let i = 1; i < data.length; i++) {
      if ((data[i] >= 0 && data[i - 1] < 0) || (data[i] < 0 && data[i - 1] >= 0)) {
        count++;
      }
    }
    return count;
  }
}

File: bci/api/python/wia_bci/core/signal_processor.py
"""WIA BCI Signal Processor."""

import math
from typing import Union

import numpy as np
from numpy.typing import NDArray

from wia_bci.types.signal import BandPowers, FrequencyBand, PowerSpectrum, FREQUENCY_BANDS


class SignalProcessor:
    """Signal processing utilities."""

    @staticmethod
    def highpass(
        data: NDArray[np.float32], cutoff: float, fs: float
    ) -> NDArray[np.float32]:
        """Apply high-pass filter (first-order IIR)."""
        result = np.zeros_like(data)
        rc = 1.0 / (2.0 * math.pi * cutoff)
        dt = 1.0 / fs
        alpha = rc / (rc + dt)

        result[0] = data[0]
        for i in range(1, len(data)):
            result[i] = alpha * (result[i - 1] + data[i] - data[i - 1])

        return result

    @staticmethod
    def lowpass(
        data: NDArray[np.float32], cutoff: float, fs: float
    ) -> NDArray[np.float32]:
        """Apply low-pass filter (first-order IIR)."""
        result = np.zeros_like(data)
        rc = 1.0 / (2.0 * math.pi * cutoff)
        dt = 1.0 / fs
        alpha = dt / (rc + dt)

        result[0] = data[0]
        for i in range(1, len(data)):
            result[i] = result[i - 1] + alpha * (data[i] - result[i - 1])

        return result

    @staticmethod
    def bandpass(
        data: NDArray[np.float32], low: float, high: float, fs: float
    ) -> NDArray[np.float32]:
        """Apply band-pass filter."""
        hp = SignalProcessor.highpass(data, low, fs)
        return SignalProcessor.lowpass(hp, high, fs)

    @staticmethod
    def notch(
        data: NDArray[np.float32], freq: float, fs: float, q: float = 30.0
    ) -> NDArray[np.float32]:
        """Apply notch filter (e.g., for 50/60 Hz line noise)."""
        result = np.zeros_like(data)
        w0 = (2.0 * math.pi * freq) / fs
        alpha = math.sin(w0) / (2.0 * q)

        b0 = 1.0
        b1 = -2.0 * math.cos(w0)
        b2 = 1.0
        a0 = 1.0 + alpha
        a1 = -2.0 * math.cos(w0)
        a2 = 1.0 - alpha

        # Normalize
        nb0 = b0 / a0
        nb1 = b1 / a0
        nb2 = b2 / a0
        na1 = a1 / a0
        na2 = a2 / a0

        x1, x2, y1, y2 = 0.0, 0.0, 0.0, 0.0

        for i in range(len(data)):
            x = data[i]
            y = nb0 * x + nb1 * x1 + nb2 * x2 - na1 * y1 - na2 * y2

            x2, x1 = x1, x
            y2, y1 = y1, y

            result[i] = y

        return result

    @staticmethod
    def psd(data: NDArray[np.float32], fs: float) -> PowerSpectrum:
        """Compute Power Spectral Density using FFT."""
        n = len(data)
        nfft = int(2 ** math.ceil(math.log2(n)))

        # Apply Hanning window
        window = 0.5 * (1.0 - np.cos(2.0 * math.pi * np.arange(n) / (n - 1)))
        windowed = np.zeros(nfft, dtype=np.float32)
        windowed[:n] = data * window

        # FFT
        fft_result = np.fft.fft(windowed)
        powers = np.abs(fft_result[: nfft // 2]) ** 2 / nfft
        frequencies = np.fft.fftfreq(nfft, 1.0 / fs)[: nfft // 2]

        return PowerSpectrum(
            frequencies=frequencies.tolist(),
            powers=powers.tolist(),
            resolution=fs / nfft,
        )

    @staticmethod
    def bandpower(
        data: NDArray[np.float32], fs: float, band: FrequencyBand
    ) -> float:
        """Calculate power in a frequency band."""
        if isinstance(band, str):
            low, high = FREQUENCY_BANDS[band]
        else:
            low, high = band

        spectrum = SignalProcessor.psd(data, fs)
        power = 0.0
        count = 0

        for freq, p in zip(spectrum.frequencies, spectrum.powers):
            if low <= freq <= high:
                power += p
                count += 1

        return power / count if count > 0 else 0.0

    @staticmethod
    def all_band_powers(data: NDArray[np.float32], fs: float) -> BandPowers:
        """Calculate all standard frequency band powers."""
        return BandPowers(
            delta=SignalProcessor.bandpower(data, fs, "delta"),
            theta=SignalProcessor.bandpower(data, fs, "theta"),
            alpha=SignalProcessor.bandpower(data, fs, "alpha"),
            beta=SignalProcessor.bandpower(data, fs, "beta"),
            gamma=SignalProcessor.bandpower(data, fs, "gamma"),
        )

    @staticmethod
    def epoch(
        data: NDArray[np.float32], start: int, end: int
    ) -> NDArray[np.float32]:
        """Extract epoch from continuous data."""
        return data[start:end].copy()

    @staticmethod
    def sliding(
        data: NDArray[np.float32], window_size: int, step: int
    ) -> list[NDArray[np.float32]]:
        """Generate sliding windows."""
        windows = []
        for start in range(0, len(data) - window_size + 1, step):
            windows.append(data[start : start + window_size].copy())
        return windows

    @staticmethod
    def detect_blinks(
        eog: NDArray[np.float32], threshold: float = 100.0
    ) -> list[int]:
        """Detect eye blinks in EOG channel."""
        blinks = []
        in_blink = False

        for i, val in enumerate(eog):
            if abs(val) > threshold and not in_blink:
                blinks.append(i)
                in_blink = True
            elif abs(val) < threshold / 2:
                in_blink = False

        return blinks

    @staticmethod
    def rms(data: NDArray[np.float32]) -> float:
        """Calculate RMS (Root Mean Square)."""
        return float(np.sqrt(np.mean(data**2)))

    @staticmethod
    def mean(data: NDArray[np.float32]) -> float:
        """Calculate mean."""
        return float(np.mean(data))

    @staticmethod
    def variance(data: NDArray[np.float32]) -> float:
        """Calculate variance."""
        return float(np.var(data))

    @staticmethod
    def zero_crossings(data: NDArray[np.float32]) -> int:
        """Count zero crossings."""
        return int(np.sum(np.diff(np.sign(data)) != 0))
